### ISSUE TITLE
Unsorted list of files and directories after reload (:reload_cwd)

### ISSUE THREAD

**Issue by nQ8hzkZw** [ranger/ranger#1508]
#### Runtime Environment

- Operating system and version: Linux Mint 19.1 MATE
- Terminal emulator and version: xterm(330)
- Python version: 3.6.7 (default, Oct 22 2018, 11:32:17) [GCC 8.2.0]
- Ranger version/commit: ranger-master 1.9.2
- Locale: en_US.UTF-8

#### Current Behavior
When any sort mode (except sort=basename and sort=natural) is enabled, the list gets sorted incorrectly after `:reload_cwd` command.
#### Expected Behavior
The list must remain sorted after reload.
Some of the sort modes (sort=extension, sort=type) should not sort the directories.

#### Steps to reproduce
1. Navigate to a directory with a large number of files and folders (for example `/etc`).
2. Enable any of the following sort modes `oe`, `ot`, `os`, `oc`, `om` or `oa`.
3. Press SHIFT+R.

#### Possible Solutions
With my limited programming skills I have not found a way to efficiently fix the problem.
With modifications below, sorting works as expected.
I tested changes with `performance_test.py`. Performance impact is ~3-5 %. This is just a workaround instead of a real fix.
I hope that someone will find and fix the root cause of the problem.

../ranger/container/directory.py:
```diff
@@ -531,6 +531,11 @@
             elif sort_func in (sort_by_basename, sort_by_basename_icase):
                 sort_func = sort_unicode_wrapper_string(sort_func)
 
+        if self.settings.sort_case_insensitive:
+            self.files_all.sort(key=sort_by_basename_icase)
+        else:
+            self.files_all.sort(key=sort_by_basename)
+
         self.files_all.sort(key=sort_func)
 
         if self.settings.sort_reverse:
```
../ranger/container/fsobject.py:
```diff
@@ -114,6 +114,7 @@
     @lazy_property
     def extension(self):
         try:
+          if not self.is_directory:
             lastdot = self.basename.rindex('.') + 1
             return self.basename[lastdot:].lower()
         except ValueError:
@@ -239,6 +240,7 @@
     @property
     def mimetype(self):
         try:
+          if not self.is_directory:
             return self._mimetype
         except AttributeError:
             self.set_mimetype()
```

**Comment by toonn** [comment]
Could you describe what's changing? I suspect you're seeing the difference between "dir sorted with on -> dir sorted by extension" and after a reload "dir sorted as returned by the OS -> dir sorted by extension." Because the sorts are stable you see the previous sort shining through as it were. So you're just seeing the result of sorting by extension starting from a different base order. (This is just a hypothesis but you should be able to test it. Start ranger with the natural sort default, change to sort by extension, note the order. Reload and note the order again. Then start ranger, changing the sort order to by extension in `rc.conf`.)

**Comment by nQ8hzkZw** [comment]
I see two unexpected changes after reload:
1. List of directories looks like randomized.
2. Files are still sorted by extension, but their names are sorted in reverse order.

<details>
  <summary>01_sort order set to natural in rc.conf</summary>

```
bin
boot
cdrom
dev
etc
home
lib
lib64
media
mnt
opt
proc
root
run
sbin
srv
sys
tmp
usr
var
initrd.img
initrd.img.old
vmlinuz
vmlinuz.old
```

</details>
<details>
  <summary>02_changed to sort by extension (after press "oe")</summary>

```
bin
boot
cdrom
dev
etc
home
lib
lib64
media
mnt
opt
proc
root
run
sbin
srv
sys
tmp
usr
var
vmlinuz
initrd.img
initrd.img.old
vmlinuz.old
```

</details>

</details>

<details>
  <summary>03_after reload (SHIFT+R)</summary>

```
dev
etc
lib
media
lib64
run
sys
bin
srv
boot
mnt
cdrom
home
sbin
opt
var
proc
tmp
usr
root
vmlinuz
initrd.img
vmlinuz.old
initrd.img.old
```

</details>

<details>
  <summary>04_sort order set to extension in rc.conf (get the same list as 03)</summary>

```
dev
etc
lib
media
lib64
run
sys
bin
srv
boot
mnt
cdrom
home
sbin
opt
var
proc
tmp
usr
root
vmlinuz
initrd.img
vmlinuz.old
initrd.img.old
```

</details>

**Comment by toonn** [comment]
This confirms my hypothesis. You'll note that the O2 is just O1 except for the files with an extension, those are resorted according to the new order.

O3 and O4 being the same suggests that both are "system order" resorted by extension. Further reloads in the O3 case shouldn't change that order again.

This isn't actually a bug but kind of a hidden feature. Some people are exploiting it to get sorting with subsorts. For example, natural first then extension means you get a sort by extension subsorted naturally. Atime then extension would be a sort by extension subsorted by atime, etc.

*If* we decide this is a bad thing. We'd have to provide a different way to achieve subsorting at the same time as "fixing" this.

**Comment by nQ8hzkZw** [comment]
Thanks for the explanation.
Now I see that my workaround brokes subsorting capabilities.
I personally prefer to sort files by extension and directories by natural sort.
It seems, at the moment there is no easy way to achieve this.

**Comment by toonn** [comment]
That's true. You're basically stuck doing `onoe` all the time while yelling "Oh, nooo!" ; )

I'll label this as an enhancement asking for a way to specify subsorts. I think there's already an issue open on that so if I find it I'll close this as a duplicate.
